drt_1d.py

```diff
@@ -23,106 +48,177 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+    
+    logging.debug(f"Calculating K_trans from {start_position} to {end_position}")
+    
+    # Since transport is symmetric, it does not matter which direction is traveled. Let's set it to travel left to right.
+    if start_position > end_position:
+        start_position, end_position = end_position, start_position
+    
+    start_voxel = voxels[0]
+    end_voxel = voxels[-1]
+    
+    # NOTE: assume equal voxel size for now. Future implementations will have variable voxel sizes.
+    voxel_size = start_voxel.voxel_size
+    sigma_s = [voxel.scattering_cross_section for voxel in voxels]
     
-    if voxel_size <= 0:
-        raise ValueError("Voxel size must be positive and non-zero.")
-    if ri < 0 or ri >= len(sigma_s) or rj < 0 or rj >= len(sigma_s):
-        raise ValueError("Voxel indices must be non-negative and within the range of the sigma_s_matrix.")
+    tau = 0.0
     
-    min_index = min(ri, rj)
-    max_index = max(ri, rj)
-    distance = abs(ri - rj) * voxel_size
-    tau = np.sum(sigma_s[min_index:max_index]) * voxel_size
-    streaming_operator = np.exp(-tau) / (4 * np.pi * distance**2)
-    logging.debug(f"Calculated K_trans: {streaming_operator}")
+    if start_voxel.index == end_voxel.index:
+        distance = end_position - start_position
+        assert len(sigma_s) == 1
+        tau = start_voxel.scattering_cross_section * distance
+        logging.debug(f"Same voxel: tau = {tau}")
+    else:
+        left_distance = (start_voxel.index + 1) * voxel_size - start_position
+        right_distance = end_position - (end_voxel.index*voxel_size)
+        
+        path = [left_distance] + [voxel_size] * (end_voxel.index - start_voxel.index - 1) + [right_distance]
+        tau = np.dot(sigma_s, path)
+        logging.debug(f"Different voxels: tau = {tau}")
+    
+    streaming_operator = (1/2) * sc.exp1(tau)
+    logging.debug(f"Calculated streaming operator: {streaming_operator}")
     return streaming_operator
 
-def construct_transition_matrix(num_voxels:int, voxel_size:float, sigma_s:np.ndarray):
+def construct_transition_matrix(voxels:list[Voxel]) -> np.ndarray:
     """
     Construct the transition matrix.
 
     Parameters:
-    - num_voxels (int): Number of spatial elements (voxels).
-    - voxel_size (float): Size of each voxel.
-    - sigma_s (numpy.ndarray): Vector of scattering cross-sections.
+    - voxels (list[Voxel]): List of Voxel objects.
 
     Returns:
     - numpy.ndarray: Transition matrix.
     """
-    logging.debug(f"Constructing transition matrix with {num_voxels} voxels and voxel size {voxel_size}")
-  
-    if num_voxels <= 0:
-        raise ValueError("Number of voxels must be positive and non-zero.")
-    if voxel_size <= 0:
-        raise ValueError("Voxel size must be positive and non-zero.")
-        
+    # NOTE: Assume sigma_s is still voxel based (will change with continuous geometry)
+    
+    if not voxels:
```

Comment:
I appreciate that this method could be called from various places, but this is now the 3rd time you've validated this data.  It leads to a lot of bloat.